pkg/gitproviders/provider_user.go

```diff
@@ -0,0 +1,225 @@
+	userRepo, err := p.provider.UserRepositories().Get(ctx, userRef)
+
+	if err != nil {
+		return fmt.Errorf("error getting user repo reference for owner %s, repo %s, %s ", owner, repoName, err)
+	}
+
+	fmt.Println("uploading deploy key")
+
+	_, err = userRepo.DeployKeys().Create(ctx, deployKeyInfo)
+	if err != nil {
+		return fmt.Errorf("error uploading deploy key %s", err)
+	}
+
+	if err = utils.WaitUntil(os.Stdout, time.Second, defaultTimeout, func() error {
+		_, err = userRepo.DeployKeys().Get(ctx, deployKeyName)
+		return err
+	}); err != nil {
+		return fmt.Errorf("error verifying deploy key %s existance for repo %s. %s", deployKeyName, repoName, err)
+	}
+
+	return nil
+}
+
+func (p userGitProvider) GetDefaultBranch(url string) (string, error) {
+	repoInfoRef, err := p.getRepoInfoFromUrl(url)
+	if err != nil {
+		return "main", err
+	}
+
+	return *repoInfoRef.DefaultBranch, nil
+}
+
+func (p userGitProvider) GetRepoVisibility(url string) (*gitprovider.RepositoryVisibility, error) {
+	repoInfoRef, err := p.getRepoInfoFromUrl(url)
+	if err != nil {
+		return nil, err
+	}
+
+	return repoInfoRef.Visibility, nil
+}
+
+func (p userGitProvider) getRepoInfoFromUrl(repoUrl string) (*gitprovider.RepositoryInfo, error) {
+	normalizedUrl, err := NewNormalizedRepoURL(repoUrl)
+	if err != nil {
+		return nil, err
+	}
+
+	repo, err := p.getUserRepo(normalizedUrl.Owner(), normalizedUrl.RepositoryName())
+	if err != nil {
+		return nil, err
+	}
+
+	repoInfo := repo.Get()
+
+	return &repoInfo, nil
+}
+
+func (p userGitProvider) getUserRepo(user string, repoName string) (gitprovider.UserRepository, error) {
+	ctx := context.Background()
+	defer ctx.Done()
+
+	userRepoRef := NewUserRepositoryRef(p.domain, user, repoName)
+
+	repo, err := p.provider.UserRepositories().Get(ctx, userRepoRef)
+	if err != nil {
+		return nil, fmt.Errorf("error getting user repository %w", err)
+	}
+
+	return repo, nil
+}
+
+func (p userGitProvider) CreatePullRequest(owner string, repoName string, targetBranch string, newBranch string, files []gitprovider.CommitFile, commitMsg string, prTitle string, prDescription string) (gitprovider.PullRequest, error) {
+	pr, err := p.createPullRequestToUserRepo(owner, repoName, targetBranch, newBranch, files, commitMsg, prTitle, prDescription)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create pull request: %w", err)
+	}
+
+	return pr, nil
+}
+
+func (p userGitProvider) createPullRequestToUserRepo(owner string, repoName string, targetBranch string, newBranch string, files []gitprovider.CommitFile, commitMessage string, prTitle string, prDescription string) (gitprovider.PullRequest, error) {
+	ctx := context.Background()
+	userRepoRef := NewUserRepositoryRef(p.GetProviderDomain(), owner, repoName)
+
+	ur, err := p.provider.UserRepositories().Get(ctx, userRepoRef)
+	if err != nil {
+		return nil, fmt.Errorf("error getting info for repo [%s] err [%s]", userRepoRef.String(), err)
+	}
+
+	if targetBranch == "" {
```

Comment:
The interface for `gitproviders.UserRespository` is embedded in `gitproviders.OrgRepository`.  Can you just create a function like below that can be used by both `orgGitProvider` and `userGitProvider` replacing the duplicate logic?

```
func createPullRequest(repo gitprovider.UserRepository, ctx context.Context, files []gitprovider.CommitFile, newBranch, targetBranch, commitMessage, prTitle, prDescription, repRef string) (gitprovider.PullRequest, error) {
	if targetBranch == "" {
		targetBranch = *repo.Get().DefaultBranch
	}

	commits, err := repo.Commits().ListPage(ctx, targetBranch, 1, 0)
	if err != nil {
		return nil, fmt.Errorf("error getting commits for repo [%s] err [%s]", repRef, err)
	}

	if len(commits) == 0 {
		return nil, fmt.Errorf("no commits on the target branch: %s", targetBranch)
	}

	latestCommit := commits[0]

	if err := repo.Branches().Create(ctx, newBranch, latestCommit.Get().Sha); err != nil {
		return nil, fmt.Errorf("error creating branch [%s] for repo [%s] err [%s]", newBranch, repRef, err)
	}

	if _, err := repo.Commits().Create(ctx, newBranch, commitMessage, files); err != nil {
		return nil, fmt.Errorf("error creating commit for branch [%s] for repo [%s] err [%s]", newBranch, repRef, err)
	}

	pr, err := repo.PullRequests().Create(ctx, prTitle, newBranch, targetBranch, prDescription)
	if err != nil {
		return nil, fmt.Errorf("error creating pull request [%s] for branch [%s] for repo [%s] err [%s]", prTitle, newBranch, repRef, err)
	}

	return pr, nil
}
```

Then you can call this function from each function:

```
func (p orgGitProvider) createPullRequestToOrgRepo(owner string, repoName string, targetBranch string, newBranch string, files []gitprovider.CommitFile, commitMessage string, prTitle string, prDescription string) (gitprovider.PullRequest, error) {
	ctx := context.Background()
	orgRepRef := NewOrgRepositoryRef(p.GetProviderDomain(), owner, repoName)

	repo, err := p.provider.OrgRepositories().Get(ctx, orgRepRef)
	if err != nil {
		return nil, fmt.Errorf("error getting info for repo [%s] err [%s]", orgRepRef.String(), err)
	}

	return createPullRequest(repo, ctx, files, newBranch, targetBranch, commitMessage, prTitle, prDescription, orgRepRef.String())
}
```

This will eliminate a lot of duplicate logic across both implementations.  The only time that org should diverge is if we happen to need to make a call to its unique function `TeamAccess`.

Reply:
fair enough, tbh honest I didn't even try to remove any duplications on the code at all to minimize the chance of introducing a bug I decided to not change the methods besides moving code around.

But I think it might be a good time to do it.

Reply:
Good catch noticing the interface for `gitproviders.UserRespository` is embedded in `gitproviders.OrgRepository` that makes everything easier